crates/dht/src/hash.rs

```diff
@@ -0,0 +1,443 @@
+//! Partition of the hash space.
+//!
+//! The space of possible hashes is mapped to a 32-bit location. See [::kitsune2_api::id::Id::loc] for
+//! more information about this. Each agent is responsible for storing and serving some part of
+//! the hash space. This module provides a structure that partitions the hash space by some factor.
+//!
+//! The factor must be chosen up front and cannot be changed. The factor determines the number of
+//! partitions to create. A factor of 32 will create a single partition. Each unit decrease from 32
+//! will double the number of partitions. A good default for a network that is expected to grow to
+//! a few thousand nodes is 21. This will create 2^11 partitions, which is 2048. Given that data is
+//! also stored redundantly, say 10 times, this is means that 20,000 nodes are needed for the data
+//! to be maximally distributed. With higher replication, say 50 times, and a hash factor of 20,
+//! over 200,000 nodes are needed for maximal distribution.
```

Comment:
Not saying no to this. Still digesting. Consider the rest of this message musings:

Part of the hope of u32 was that if an app really took off we'd still be a long way off from every peer having an arc-length of 1 and still covering the whole space. I.e. room to grow.

A subset of happs being written will know they won't ever hit that scale, and may be fine with a smaller bound. In that case, maybe we should use a u16, and not have a factor at all?

Another subset of happs being written will probably never hit that scale, but they'll have the hope of it. In which case, maybe we should go ahead and use a u32 and not have a factor at all, and figure out some way to compress empty locations within a time slice the way we did with empty time slices?

Reply:
As you say, the agents pick the number of chunks they want to cover. That is the scaling metric. The factor here boils down to a choice (presumably at app bundling time) of how much network overhead the developer is willing to accept in exchange for being able to shard more if their app is well used.

u32 was chosen before the time partitioning was a thing... but while the individual op count within any time-bounded arc range will probably be kinda small for most apps, we do still have to consider the bandwidth requirement for full-syncing new nodes into the network.

If an app had a total history burden of 1 Terabyte, if you had an arc size of 1 out of u16, you would only have to store 16 Megabytes of that. So it seems (to me) that we would never need to shard beyond 1:65536.

The same calculation for a u8 yields individual nodes needing to be responsible for 4 Gigabytes. While that doesn't seem entirely unreasonable, it does feel on the edge of maybe there are use-cases that might want to shard beyond that.

Correct me if I'm wrong about the following assumptions:

- We don't care about the network overhead in a full-sync operation. It will be insignificant compared to the op data that needs to be transferred.
- We don't care about the network overhead in a these-nodes-have-last-seen-timestamps of each other, because there is essentially no overhead there, we just send the hashes we've seen since then.
- We DO care about the network overhead in the needing-to-compare-hashes-of-partitions case... especially if there are a lot of partitions--say 65536 per time-slice because the devs are hoping their app will grow crazy. Because with 32-byte hashes, that is already 2M per time slice.

Reply:
I expected this to be a controversial point so musing is good :)

Does full-sync here refer to new agents joining or all agents holding everything? 
I think in the new agent case it's a yes, there would be some overhead to communicate what the top hashes should be to aim for but that should be minimal compared with the data itself.
In the case of agents holding everything, communicating top hashes at a point in time can be combined right the way down to one 32-bit hash as a summary. The problem is when there's a mismatch for some reason, which could quite easily happen if an op gets lost. If we have a really large number of hash partitions and are managing time slices for each then we have to send at least a top level time hash for each hash partition... which is already thousands of hashes at the scale I'm talking about in these docs.
To make that work, I think the network does need to freeze time slices after a while and stop trying to sync them with peers. Ideally a few weeks. At least that way we aren't going to keep going through that loop in gossip of discovering some ancient mismatch and having to repeatedly send lots of hashes in each gossip round.
So what I'm talking about here is the third point I think, about how many partitions we have to compare hashes over.
So that's some first thoughts from me on the topic of "how much sharding should we support". I think once it goes beyond a certain point, full sync nodes are going to be comparing a lot. If you get right down to agents only holding once chunk then it's very cheap. So there is a bit of a tradeoff there depending on how you expect the network to behave. But if you gain enough data quickly for the network to shard, then where are the limits...

Taking something like a 1 Terrabyte network. How much of that do I care about? If I'm storing 4GB but I'm also a participating node that isn't just providing storage. I need to be able to get and cache whatever data I'm interacting with. So say it's something like a social app and I produce something like 200Mb worth of text and images but then I have a couple of hundred friends doing the same. The size of the cache to participate is going to grow rapidly too. Sure I don't have to provide the processing power to validate that data and I can aggressively prune my cache. But I could easily end up with 5-10Gb of data that's in active use if my friends are similarly active and we assume I'm commenting back on threads that people have posted a few weeks ago.

So I'm thinking that in a network that active, there will likely be such a cache requirement to participate that it's acceptable to have a reasonable amount of DHT data to be responsible for. 

Then if we want to hit mega scale. I'm not sure that every participant providing processing power is the right way to go. I think a deployment model where most people have a pinned 0 arc and rely on their cache is going to be the best option. Then run many machines that have 100Gb free for storage. We have to expect some validation to require gets of data that is going to be cached rather than DHT data. So let's be conservative and say that a node with 100Gb can really only offer 60Gb of valuable storage, with aggressive cache cleaning. Then I think that with this level of partitioning the app can actually be scaled to a good level.

All just musings and trying to get across the way I'm thinking about this. I'm in agreement that the same thing would probably work with a u16 and we'd just have smaller numbers to deal with but I'm mostly interested in figuring out whether this works to solve the "how the network scales" questions.

I'm also happy to sit down and do the calculations properly rather than roughly if that would add some value to the discussion. Everything here is pretty much from my head :)

Reply:
yep sorry, that was a bad use of full-sync. I meant the thing where a new node joins the data and need to catch up to everything for the idea arc size.

Reply:
Okay, I think I've identified what my concern here is. You are correct, the bit width of the location/arc bounds is irrelevant because it just affects local calculations.

Rather than making the factor configurable, I vote for hard-coding it. I'm just going to throw out a number and say 512. I assert that all small and medium scale apps will still ideally hold arc ranges of more than 1 in that 512. Beyond that, a large scale app will function just fine with everyone holding exactly 1 in 512. And you are correct, there are other mitigations that can happen if individual nodes are starting to get overwhelmed in even that scenario. And 512 is small enough that I feel like it will be manageable in terms of coordination overhead, although we should get some numbers on that.

We can always change this factor by protocol version bump. We don't want to make developers make a decision on this that they won't really understand. And if we want to get more sophisticated in later iterations, I think we could probably come up with a way to derive a consensus factor from the data in the dht, or negotiate it in connection preflights even if that just means always picking the higher partition count number.

Reply:
So, having said that, your code here shouldn't need to change at all, so I'll continue my review : )

Reply:
> To make that work, I think the network does need to freeze time slices after a while and stop trying to sync them with peers. Ideally a few weeks.

By design, there's no consensus in holochain, and by extension kitsune. The same dna hash could be partitioned on two LANs for multiple years, and then come together some day. The nodes will see each other, but if freezing is happening, they won't be able to validate ops properly.

That would certainly be a rare scenario, but it'd be awesome if we could easily support it.

Once we have the hashes in memory, it should be trivial (computationally) to build up a cross-cutting merkle-like tree of them for all arc buckets but a single time-slice. Then (the hope would be) that the older stuff wouldn't change very much, and we could exchange our top-level time slice hashes without much overhead.

No need to do this now, but if you think something like this (or something else entirely) might someday be able to solve this problem, I'd be a lot more comfortable : )

Reply:
I would hold off on assuming we're going to freeze fixed time slices. We have no data on this, and it may turn out to not be an issue to compare and modify them.

Reply:
Over time it does become an issue I believe. I am fine if we say there's no consensus by design and that means we can't actually refuse to accept updates. However, gossip will degenerate when there is a permanent historical mismatch.

Say we have 1k full time slices. What we'd like to be able to do is aggregate those and expect mismatches only in the partial time slices. We have to expect that in some gossip rounds that isn't what will happen, we're going to run into somebody who we aren't in sync with. In that case, we have to go more granular on the full time slices but we don't want to send more data than we have to, to find the problem. So we fall back to something like telescoping the full time slices into combined hashes. That's not so bad, we're still sending a reasonable amount of data at that point still. We're still relying on probability, where we expect more recent time slices to be more likely to contain a mismatch. So it's okay to roll up more older time slices because we don't think they're so likely to mismatch. Finally though, if we find a mismatch in one of these telescoped slices, we either need to send all hashes in that slice or we need to go round again and send all the top hashes for the time slices that made up that telescoped slice.

That all works fine and won't happen that often. The problem comes when we have a peer who mismatches with us at a low slice and we can't get one or more problem hashes. Then every single gossip round will degenerate down to the worst case and we'll have to send them a large batch of top hashes.

I think that assuming we can always get every op is a mistake. If the system is consensus free then we don't need to match every hash. We just need to do our best to match. At some point, surely it makes sense to protect bandwidth?

I'm fine with treating this as an edge case and not doing anything with it for now. However, I think a potential solution might look like pinning some time hashes for peer where we have a mismatch which is "ancient". Then when we compute a top hash for gossip with that peer, we substitute some full slice hashes of our own, with their value for that time slice. That trades a small amount of storage for a whole lot of bandwidth.
Sure it creates a new problem that a peer can force us to use disk space. But that can be solved too.

Reply:
Can you describe your thinking about how to compute the "top_hash"?

Reply:
> That all works fine and won't happen that often. The problem comes when we have a peer who mismatches with us at a low slice and we can't get one or more problem hashes. Then every single gossip round will degenerate down to the worst case and we'll have to send them a large batch of top hashes.

When we can't get one or more problem combined hashes, we exchange top hashes at the next lower level, right? And if that continues and we don't have any combined hashes we agree on, we start comparing individual op hashes, and since we have the same number of full time slices but don't agree on any combined hashes for any of those slices, we need to compare all op hashes of historical gossip. That is the case you're describing here, right?

I would like to know the probability of this case. It means that another peer has sufficiently gossiped around to know about all the time slices there have been since the beginning of the app, but does not have a single combined hash that matches ours, meaning there's a diff for every of the full slices. Is that what we're talking about?

Reply:
Is "top hash" the same as `PartialSlice::hash` in `/dht/src/time.rs`?

Reply:
@zippy yes

Reply:
I see, so it's actually the result of `combine_op_hashes` for both partial and full slices.   I notice that this conversation has also included the word "chunk" which appears to by synonymous with "slice" which is what's in the code.  Maybe "chunk" existed in the previous implementation?

Reply:
There's a huge amount of information in this thread now. I'm going to respond about terminology but I think we should split out the conversation about historical mismatches. It's not actually relevant to how we partition, it's relevant to the gossip process that's built on top of it later. For me the link is how granular the partitioning is. The more granular we are with time and space, the more important it is that we don't have to keep re-negotiating historical mismatches. So for this conversation let's just say that we want to keep the number of partitions as low as possible while also allowing Kitsune2 to scale appropriately.

As we iterate, the terms we're using are changing so there's going to have to be some shared adopting/dropping of terms and I hope that we can land on something consistent as this comes together. I'm trying to use the following terms:
- Time slice - Some period of time that is a right-open interval (e.g. May 5, from and including midnight to May 9th, up to but not including midnight)
- Combined hash - The reduce/fold of a set of op hashes, using an XOR over those hashes. When we're talking about a time slice that has no implied "tree" structure, then this is the term I'm using.
- DHT arc - Refers to some part of the DHT hash space. Because of how we're partitioning it can be thought of as a contiguous list of segments which are each the same size. So say we break the hash space into 4 pieces then a valid DHT arc would be 1, 2, 3 or all 4 of those pieces but 2 and 4 wouldn't be possible to represent, it would have to be 1 and 2 or 2 and 3 or 2, 3 and 4.
- Full time slice - A historical time slice that is supposedly complete, bar any network partitioning or offline nodes
- Partial time slice - A recent time slice where we expect regular updates
- Top hash - When there is some implied structure by considering some topology over the space and time. So that could be segments where we roll up full time slices within DHT arcs. It could be a disc where we combine over DHT arcs and full time slices. It could be a ring where we combine over all DHT arcs for a single partial time slice. It is intended to convey the meaning of a "summary of summaries" to compress information that we think is likely to match somebody else's view of the DHT.

Reply:
So `PartialSlice::hash` is not a top hash, it is a "combined hash" as documented and I do want to keep those terms separate. I think "chunk" is confusing and where I've used it then it was an accidental use of terminology from Kitsune 1. I mean holding a single DHT arc, at the defined unit size.

Reply:
Thanks that's really helpful.  It's really helpful if that king of terminology glossary can make it into the spec, or a readme.